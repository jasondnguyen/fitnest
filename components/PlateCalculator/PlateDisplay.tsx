import * as React from 'react';
import { StyleSheet, Image, View, Text, Dimensions } from 'react-native';
import Images from '../Layout/images';

const size_proportion = Dimensions.get('window').height * 0.1;
const font_proportion = Dimensions.get('window').height * 0.05;

interface Props {
  metric: string;
  plate: Object;
}

const selectImage = (metric: string, plate: number) => {
  const plateArray = {
    lb: {
      45: Images.lb[45],
      35: Images.lb[35],
      25: Images.lb[25],
      10: Images.lb[10],
      5: Images.lb[5],
      2.5: Images.lb[2.5],
    },
    kg: {
      20: Images.kg[20],
      15: Images.kg[15],
      10: Images.kg[10],
      5: Images.kg[5],
      2.5: Images.kg[2.5],
      1.25: Images.kg[1.25],
    },
  };

  return plateArray[metric][plate];
};

const PlateDisplay: React.FC<Props> = ({ metric, plate }) => {
  const plateValue = Object.keys(plate)[0];
  const plateAmount = Object.values(plate)[0];

  return (
    <View style={styles.container}>
      <Text style={{ fontSize: font_proportion }}>
        <Image
          source={selectImage(metric, plateValue)}
          style={{ height: size_proportion, width: size_proportion }}
        />
        x{plateAmount}
      </Text>
    </View>
  );
};

export default PlateDisplay;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
  },
});
