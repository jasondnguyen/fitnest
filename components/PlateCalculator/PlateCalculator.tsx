import * as React from 'react';
import { TextInput, ToggleButton, Button } from 'react-native-paper';
import { StyleSheet, Alert, Dimensions, View } from 'react-native';
import PlateDisplay from './PlateDisplay';

const width_proportion = '80%';
const icon_proportion = Dimensions.get('window').width * 0.1;

const calculateTotalPlates = (weight: number, metric: string) => {
  try {
    if (weight % 5 != 0 || weight < 45) {
      Alert.alert('Error', 'Please enter a valid weight');
    }

    const plates: { [key: string]: Array<number> } = {
      lb: [45, 35, 25, 10, 5, 2.5],
      kg: [20, 15, 10, 5, 2.5, 1.25],
    };

    const neededPlates: Array<Object> = [];
    weight = (weight - 45) / 2;

    for (let plate of plates[metric]) {
      const neededPlate = Math.floor(weight / plate);
      neededPlates.push({ [`${plate}`]: neededPlate * 2 });
      weight -= neededPlate * plate;
    }

    return neededPlates;
  } catch (err) {
    console.log(err);
  }
};

const PlateCalculator = () => {
  const [weight, setWeight] = React.useState('');
  const [metric, setMetric] = React.useState('lb');
  const [plates, setPlates] = React.useState([] as any);

  return (
    <View style={styles.root}>
      <TextInput
        label='Weight'
        value={weight}
        onChangeText={text => setWeight(text)}
        placeholder='145'
        style={styles.weightInput}
      />
      <ToggleButton.Row
        onValueChange={chosenMetric => {
          if (chosenMetric) {
            setMetric(chosenMetric);
            if (chosenMetric !== metric) {
              setPlates([]);
            }
          }
        }}
        value={metric}
        style={styles.weightRow}
      >
        <ToggleButton
          icon='weight-pound'
          value='lb'
          style={{ width: icon_proportion }}
        />
        <ToggleButton
          icon='weight-kilogram'
          value='kg'
          style={{ width: icon_proportion }}
        />
      </ToggleButton.Row>
      <Button
        mode='contained'
        onPress={() =>
          setPlates(calculateTotalPlates(parseInt(weight), metric))
        }
      >
        Calculate Plates
      </Button>
      {plates &&
        plates.map((plate: Object, index: number) => (
          <PlateDisplay key={index} plate={plate} metric={metric} />
        ))}
    </View>
  );
};

export default PlateCalculator;

const styles = StyleSheet.create({
  root: {
    flex: 1,
  },
  weightInput: {
    width: width_proportion,
    height: 44,
  },
  weightRow: {
    left: width_proportion,
    position: 'absolute',
  },
});
