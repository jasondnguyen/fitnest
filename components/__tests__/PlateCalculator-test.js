import * as React from 'react';
import renderer from 'react-test-renderer';

import PlateCalculator from '../PlateCalculator/PlateCalculator';
import calculateTotalPlates from '../PlateCalculator/CalculatorLogic';

it(`renders correctly`, () => {
  const calculator = renderer
    .create(<PlateCalculator>Snapshot test!</PlateCalculator>)
    .toJSON();

  expect(calculator).toMatchSnapshot();
});

it('calculates correct plates in pounds', () => {
  expect(calculateTotalPlates(135, 'lb')).toBe([
    { bar: 1 },
    { 45: 2 },
    { 35: 0 },
    { 25: 0 },
    { 10: 0 },
    { 5: 0 },
    { 2.5: 0 },
  ]);
});
