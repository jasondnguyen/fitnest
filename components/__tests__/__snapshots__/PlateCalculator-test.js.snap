// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgb(231, 231, 231)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        Object {
          "width": "80%",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.26)",
          "bottom": 0,
          "height": 2,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": Array [
            Object {
              "scaleY": 0.5,
            },
          ],
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          Object {
            "minHeight": 43,
          },
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <Text
          numberOfLines={1}
          onLayout={[Function]}
          style={
            Object {
              "color": "#6200ee",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "left": 0,
              "opacity": 0,
              "paddingLeft": 12,
              "paddingRight": 12,
              "position": "absolute",
              "textAlign": "left",
              "top": 21,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
                Object {
                  "scale": 1,
                },
              ],
              "writingDirection": "ltr",
            }
          }
        >
          Weight
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.54)",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "left": 0,
              "opacity": 0,
              "paddingLeft": 12,
              "paddingRight": 12,
              "position": "absolute",
              "textAlign": "left",
              "top": 21,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
                Object {
                  "scale": 1,
                },
              ],
              "writingDirection": "ltr",
            }
          }
        >
          Weight
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        editable={true}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder=""
        placeholderTextColor="rgba(0, 0, 0, 0.54)"
        rejectResponderTermination={true}
        selectionColor="#6200ee"
        style={
          Array [
            Object {
              "flexGrow": 1,
              "margin": 0,
              "zIndex": 1,
            },
            Object {
              "paddingLeft": 12,
              "paddingRight": 12,
            },
            Object {
              "height": 43,
            },
            Object {
              "paddingBottom": 4,
              "paddingTop": 24,
            },
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            Array [
              Object {},
            ],
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Object {
          "left": "80%",
          "position": "absolute",
        },
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
          "selected": true,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          false,
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            undefined,
            Array [
              Object {
                "height": 42,
                "margin": 0,
                "width": 42,
              },
              Object {
                "backgroundColor": "rgba(0, 0, 0, .08)",
                "borderColor": "rgba(0, 0, 0, 0.29)",
                "borderRadius": 4,
              },
              Array [
                Object {
                  "borderWidth": 0.5,
                },
                Object {
                  "borderBottomRightRadius": 0,
                  "borderTopRightRadius": 0,
                },
                undefined,
              ],
            ],
          ],
        ]
      }
    >
      <View>
        <Text
          accessibilityElementsHidden={true}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          selectable={false}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "color": "#000000",
                "fontSize": 24,
              },
            ]
          }
        >
          □
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
          "selected": false,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          false,
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            undefined,
            Array [
              Object {
                "height": 42,
                "margin": 0,
                "width": 42,
              },
              Object {
                "backgroundColor": "transparent",
                "borderColor": "rgba(0, 0, 0, 0.29)",
                "borderRadius": 4,
              },
              Array [
                Object {
                  "borderWidth": 0.5,
                },
                Object {
                  "borderBottomLeftRadius": 0,
                  "borderLeftWidth": 0,
                  "borderTopLeftRadius": 0,
                },
                undefined,
              ],
            ],
          ],
        ]
      }
    >
      <View>
        <Text
          accessibilityElementsHidden={true}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          selectable={false}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "color": "#000000",
                "fontSize": 24,
              },
            ]
          }
        >
          □
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#6200ee",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 2,
        "minWidth": 64,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0.75,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 1.5,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#ffffff",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          Calculate Plates
        </Text>
      </View>
    </View>
  </View>
</View>
`;
